std/range/primitives.d

```diff
@@ -1986,6 +1992,140 @@ ElementType!R moveAt(R, I)(R r, I i) if (isIntegral!I)
     }
 }
 
+private enum int hasNothrowPopFront(Range) = (functionAttributes!(popFrontN!Range)
+                                               & FunctionAttribute.nothrow_);
+
+/**
+Pops the first element from a range or array $(D r) and returns it.
+It combines $(D front) and $(D popFront) in one method.
+
+Params:
+    r = range to consume the first element from
+
+Warning:
+    If a transient range is used, you should `dup`licate it to avoid it from
+    being overwritten.
+    Using shift with `popFront` that might `throw` could lead to a loss of the
+    `front` element and thus it is prohibited.
+
+Returns:
+    Next element of the input range
+ */
+auto shift(Range)(ref Range r)
+if (isInputRange!Range && hasNothrowPopFront!Range)
+in
+{
+    assert(!r.empty, "empty range");
+}
+body
+{
+    auto e = r.front;
+    r.popFront();
+    return e;
+}
+
+///
+@safe pure unittest
```

Comment:
Thanks.
Yes, but it shows an interesting (for newbies) effect: only the slice gets modified and not the original array, which IIRC is not the case in other languages like JS.
Maybe it's worth adding `assert (arr.equals(4.iota))` at the end.